text/3529-cargo-path-bases.md

```diff
@@ -0,0 +1,395 @@
+`Cargo.toml` files, you can define it as a "base" path in
+`~/.cargo/config.toml`:
+
+```toml
+[base-paths]
+dev = "/home/user/dev/rust/libraries/"
+```
+
+Now, you can specify a path dependency on a library `foo` in that
+directory in your `Cargo.toml` using
+
+```toml
+[dependencies]
+foo = { path = "foo", base = "dev" }
+```
+
+Like with other path dependencies, keep in mind that both the base _and_
+the path must exist on any other host where you want to use the same
+`Cargo.toml` to build your project.
+
+# Reference-level explanation
+[reference-level-explanation]: #reference-level-explanation
+
+## Specifying Dependencies
+
+### Base Paths
+
+A `path` dependency may optionally specify a base path by setting the `base` key
+to the name of a base path from the `[base-paths]` table in either the
+[manifest](https://doc.rust-lang.org/cargo/reference/manifest.html) or
+[configuration](https://doc.rust-lang.org/cargo/reference/config.html#base-paths)
+or one of the [built-in base paths](#built-in-base-paths). The value of that
+base path is prepended to the `path` value to produce the actual location where
+Cargo will look for the dependency.
+
+For example, if the Cargo.toml contains:
+
+```toml
+[dependencies]
+foo = { path = "foo", base = "dev" }
+```
+
+Given a `[base-paths]` table in the configuration that contains:
+
+```toml
+[base-paths]
+dev = "/home/user/dev/rust/libraries/"
+```
+
+This will produce a `path` dependency `foo` located at
+`/home/user/dev/rust/libraries/foo`.
+
+If the base path is not found in any `[base-paths]` table or one of the built-in
+base paths then Cargo will generate an error.
+
+If the name of a base path is specified in both the manifest and configuration,
+then the value in the manifest is preferred.
+
+The name of a base path must use only [alphanumeric](https://doc.rust-lang.org/std/primitive.char.html#method.is_alphanumeric)
+characters or `-` or `_`, must start with an [alphabetic](https://doc.rust-lang.org/std/primitive.char.html#method.is_alphabetic)
+character, and must not be empty.
+
+#### Built-in base paths
+
+Cargo provides implicit base paths that can be used without the need to specify
+them in a `[base-paths]` table.
+
+* `workspace` - If a project is [a workspace or workspace member](https://doc.rust-lang.org/cargo/reference/workspaces.html)
+then this base path is defined as the path to the directory containing the root
+Cargo.toml of the workspace.
+
+If one of these built-in base paths is also specified in the manifest or
+configuration, then that value is preferred over the built-in value. This allows
+Cargo to introduce additional built-in paths in the future without compatibility
+issues, as any existing uses will shadow Cargo's built-in name.
+
+## The Manifest Format
+
+[`[base-paths]`](https://doc.rust-lang.org/cargo/reference/specifying-dependencies.html#base-paths) - Base paths for path dependencies.
+
+## Configuration
+
+`[base-paths]`

```

Comment:
Does this table name need to be more specific in some way.  In either config or manifest, `base-paths` has little meaning.

Reply:
While this seems like a very minor tweak, I wonder if even `path-bases` would be better....

Reply:
What about `dependency-base-paths`? or `path-dependency-bases`?

Reply:
@epage wrote:
> While this seems like a very minor tweak, I wonder if even `path-bases` would be better....

That does seem like a bit of an improvement, yeah. That makes it clear "these are the bases for `path`s", rather than the possible implication of "these are the paths of bases, but bases of what?"

Reply:
Switched to `path-bases`